native/src/main/scala/sqs4s/auth/CredProvider.scala

```diff
@@ -0,0 +1,159 @@
+      RetriableServerError.pure[F].widen
+    } else {
+      UnknownAuthError(resp.status).pure[F].widen
+    }
+  }
+
+  val refresh: F[Credential] = {
+    val newCred =
+      for {
+        token <- client.expectOr[String](PUT(tokenEndpoint))(onError)
+        creds <- client.expectOr[List[Credential]](
+          GET(
+            credsEndpoint,
+            Header(ttlHeader, ttl.toSeconds.toString),
+            Header(tokenHeader, token)
+          )
+        )(onError)
+        cr <- Sync[F].fromOption(
+          creds.headOption,
+          NoInstanceProfileCredentialFound
+        )
+      } yield cr
+    for {
+      now <- Clock[F].realTime(TimeUnit.MILLISECONDS)
+      optCred <- ref.get
+      refreshed <- optCred match {
+        case Some(cred) if cred.expiration.getEpochSecond > now =>
+          cred.pure[F]
+
+        case _ =>
+          newCred.flatMap(cred => ref.set(cred.some).as(cred))
+      }
+    } yield refreshed
+  }
+
+  val refreshAsync: Stream[F, Unit] = {
+    val constRefresh = Stream.repeatEval(refresh).metered(
+      ttl - refreshBefore
+    ).compile.drain
+    Stream.retry(
+      constRefresh,
+      0.second,
+      _ => 1.second,
+      10,
+      {
+        case RetriableServerError => true
+        case _ => false
+      }
+    )
+  }
+
+  val accessKey: F[String] =
+    ref.get.flatMap {
+      case Some(cred) => cred.accessKeyId.pure[F]
+      case None => refresh.map(_.accessKeyId)
+    }
+
+  val secretKey: F[String] =
+    ref.get.flatMap {
+      case Some(cred) => cred.secretAccessKey.pure[F]
+      case None => refresh.map(_.secretAccessKey)
+    }
+
+  val sessionToken: F[String] =
+    ref.get.flatMap {
+      case Some(cred) => cred.token.pure[F]
+      case None => refresh.map(_.token)
+    }
+}
+
+object InstanceProfileCredProvider {
+
+  def instance[F[_]: Sync: Clock: Timer](
+    client: Client[F],
+    ttl: FiniteDuration = 6.hours,
+    refreshBefore: FiniteDuration = 1.minute
+  ): F[InstanceProfileCredProvider[F]] =
+    Ref.of[F, Option[Credential]](None).map(cred =>
+      new InstanceProfileCredProvider[F](client, cred, ttl, refreshBefore))
+
+  /**
+    * Create an instance of InstanceProfileCredProvider but start a refresh
+    * stream in another thread which rotate credential every ttlDefault - 60 seconds
+    * @param client http client
+    * @param ttl token's ttl (token's rotation interval)
+    * @tparam F a concurrent effect so that token can be refresh in another thread
+    * @return an instance of InstanceProfileCredProvider
+    */
+  def asyncRefreshInstance[F[_]: Concurrent: Clock: Timer](
+    client: Client[F],
+    ttl: FiniteDuration = 6.hours,
+    refreshBefore: FiniteDuration = 1.minute
+  ): F[InstanceProfileCredProvider[F]] =
+    Ref.of[F, Option[Credential]](None).flatMap { cred =>
+      val provider =
+        new InstanceProfileCredProvider[F](client, cred, ttl, refreshBefore)
+      provider.refreshAsync.compile.drain.start.as(provider)
```

Comment:
Still not 100% happy with this line here. Basically what it does it run a fs2 Stream in a `Fiber` so it is a bit like fire and forget. What that stream does is to run `refresh` to rotate the cred periodically.

Reply:
Do we really need that? Isn't it easier _(even better)_ to just run `refresh` if we get an **Unauthorized**?
If after the `refresh` we get another **Unauthorized** then fail with an error.

Reply:
that is what I'm already doing [here](https://github.com/d2a4u/sqs4s/pull/128/files#diff-3b5fb231906d76041d851d15b053f5f1R40). There are 2 ways to use this `InstanceProfileCredProvider`, 1st one is like you described, which already covered, 2nd one is using this `asyncRefresh`, again, this is something that the Java SDK has and I was thinking it is actually pretty cool feature to have, to reduce the number of 401 error and retries. Maybe I should make `refresh` returns `Unit` instead.

Reply:
> that is what I'm already doing here.

Yeah, I just read my comment again and I realized that it gave the impression that I was saying you didn't do that, sorry.
My intention was to say that should be enough.

> I was thinking it is actually pretty cool feature to have, to reduce the number of 401 error and retries.

I also think it is a cool feature.
However, I believe that one makes more sense for other services that require HA and low throughput.
SQS is considered for simple applications were a delay shouldn't be a problem _(especially considering we have to pull over some rate instead of getting a push)_. So, my humble opinion is that the retry on failure should be more than enough.

_Neverthless, just because I as you want to make cool software, take a look to [this](https://www.reddit.com/r/scala/comments/ho6hr8/fp_approach_to_manage_the_lifecycle_of_an_api/fxgp6j9?utm_source=share&utm_medium=web2x) basically the idea would be that the `async` one returns a resource which under the hood uses a **Signal**. Also, it would be good if the `async` one overrides `refresh` as a no-op_.

> Maybe I should make refresh returns Unit instead.

+1